File: backend/app/schemas/reservation.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator
from app.schemas.listing import ListingResponse


class ReservationCreate(BaseModel):
    listing_id: str
    check_in: date
    check_out: date
    guests: int = Field(default=1, gt=0)

    @model_validator(mode="after")
    def validate_dates(self) -> "ReservationCreate":
        if self.check_out <= self.check_in:
            raise ValueError("check_out date must be strictly after check_in date")
        return self


class ReservationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    listing_id: str
    guest_id: str
    check_in: date
    check_out: date
    guests: int
    total_price: float
    created_at: datetime
    listing: Optional[ListingResponse] = None

File: frontend/src/types/reservation.ts
import type { Listing } from "./listing";
import type { User } from "./user";

export interface Reservation {
  id: string;
  listing_id: string;
  guest_id: string;
  check_in: string; // ISO date string "YYYY-MM-DD"
  check_out: string; // ISO date string "YYYY-MM-DD"
  guests: number;
  total_price: number;
  created_at: string;
  // Populated by the backend when fetching with joins
  listing?: Listing;
  guest?: User;
}

export interface CreateReservationPayload {
  listing_id: string;
  check_in: string;
  check_out: string;
  guests: number;
}
